In [1]:
import pandas as pd
import numpy as np

### Data Preprocessing

In [2]:
data = pd.read_csv(r"C:\Users\Triloki-Gupta\Downloads\pvdiagnostic\Test-A\20190207125454041.csv")
print(data.shape)
data.head()

(438, 2)


Name       STRING
0               Date  07-02-2019 
1               Time    12:54:54 
2  Temperature[degC]   47.652863 
3    Voltge Range[V]        1000 
4   Current Range[A]          30

In [183]:
col = []
for i in range(35):
    col.append(data.loc[i,"Name"])

In [184]:
new_data = pd.DataFrame(columns=col, dtype="int")
new_data

Empty DataFrame
Columns: [Date, Time, Temperature[degC], Voltge Range[V], Current Range[A], Sampling Pitch[uS], Sensor, Caribration Constant (Er cal)[mV/kW/m^2], PV Device Area[m^2], Current Temp. Coef.[A/degC], Voltage Temp. Coef.[V/degC], Series Resistance(Rs)[ohm], Curve Correction Factor (Kappa)[ohm/degC], Series Modules, Parallel Modules, Solar Irradiance (Er)[W/m^2], PV Device Temp.[degC], Isc[A], Voc[V], Pm[W], Ipm[A], Vpm[V], Eta[%], FF, Er (stc)[W/m^2], T (stc)[degC], Isc (stc)[A], Voc (stc)[V], Pm (stc)[W], Ipm (stc)[A], Vpm (stc)[V], Eta (stc)[%], FF(stc), nan, Voltage[V]]
Index: []

[0 rows x 35 columns]

In [179]:
from os import listdir
from os.path import isfile, join
mypath = r'C:\Users\Triloki-Gupta\Downloads\pvdiagnostic\Test-A'
onlyfiles = [f for f in listdir(mypath) if isfile(join(mypath, f))]

In [217]:
for j in range (len(onlyfiles)):
    print(j)
    data = pd.read_csv(mypath+"\\"+onlyfiles[j])
#     for j in range(len(mylist)):
    for i in range(35):
        try:
#                 print(data.loc[i,"STRING"])
            new_data.loc[j,col[i]]=data.loc[i,"STRING"]
            new_data.loc[j,"Module No."]=str(onlyfiles[j].split('.')[0])

        except:
#                 print(data.loc[i,"MODULE"])
            new_data.loc[j,col[i]]=data.loc[i,"MODULE"]
            new_data.loc[j,"Module No."]=str(onlyfiles[j].split('.')[0])

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195


In [218]:
new_data.head()

Date       Time Temperature[degC] Voltge Range[V] Current Range[A]  \
0  07-02-2019   12:54:54         47.652863            1000               30    
1  16-02-2019   10:55:44         47.964508             100               10    
2  16-02-2019   12:10:24         51.697929             100               10    
3  16-02-2019   12:14:36         50.773357             100               10    
4  16-02-2019   12:15:18         50.338634             100               10    

  Sampling Pitch[uS]       Sensor Caribration Constant (Er cal)[mV/kW/m^2]  \
0                10   Pyranometer                                   52.49    
1                10   Pyranometer                                   52.49    
2                10   Pyranometer                                   52.49    
3                10   Pyranometer                                   52.49    
4                10   Pyranometer                                   52.49    

  PV Device Area[m^2] Current Temp. Coef.[A/degC]  ... Isc (stc)[A]  \
0                  1                           0   ...     8.194866   
1                  1                           0   ...     7.786469   
2                  1                           0   ...     7.765487   
3                  1                           0   ...     7.651004   
4                  1                           0   ...     7.632927   

  Voc (stc)[V]  Pm (stc)[W] Ipm (stc)[A] Vpm (stc)[V] Eta (stc)[%]  \
0   901.289707  5245.071996     7.371913   711.493992   11.583047    
1    36.603990   389.943103    10.372680    37.593285   17.987082    
2    36.628285   214.668875     5.703425    37.638593    17.02744    
3    36.129507   214.990621     5.723115    37.565314   16.785577    
4    36.317609   148.847265     3.768207    39.500818   11.905059    

      FF(stc) NaN  Voltage[V]         Module No.  
0     0.6927  NaN  Current[A]  20190207125454041  
1   0.652611  NaN  Current[A]  20190216105544030  
2   0.647208  NaN  Current[A]  20190216121024130  
3   0.653564  NaN  Current[A]  20190216121436138  
4   0.462546  NaN  Current[A]  20190216121518140  

[5 rows x 36 columns]

In [227]:
data_B = pd.read_excel(r"C:\Users\Triloki-Gupta\Downloads\pvdiagnostic\Test-B\Test-B results (dataset-B).xlsx", dtype={'Module No.':'int64'})
print(data_B.shape)
data_B.head()

(196, 2)


Module No. Test-B result
0  20190418111356171        normal
1  20190418124126297        normal
2  20190418124208299        normal
3  20190418124228300        normal
4  20190418142022001        normal

In [234]:
new_data["Module No."] = new_data["Module No."].astype("int64")

In [237]:
#merging file on the base of module no. and saving
rawData=pd.merge(new_data,data_B, on='Module No.')
# final_data.to_excel(r"C:\Users\Triloki-Gupta\Downloads\pvdiagnostic\dataWithClass.xlsx", index=False)
rawData.to_csv(r"C:\Users\Triloki-Gupta\Downloads\pvdiagnostic\dataWithClass.csv", index=False)

In [3]:
#loading withclass data
dataWithClass = pd.read_csv(r"C:\Users\Triloki-Gupta\Downloads\pvdiagnostic\dataWithClass.csv",dtype={'Module No.':'object'})

## EDA

In [4]:
#unique class/target
dataWithClass["Test-B result"].unique()

array(['defect_S_C', 'defect_C', 'defect_S', 'defect_P', 'normal',
       'defect_S_P'], dtype=object)

In [5]:
#top 3 data point
dataWithClass.head(3)
# dataWithClass.select_dtypes(include = ['object']).head(3)

Date       Time  Temperature[degC]  Voltge Range[V]  \
0  07-02-2019   12:54:54           47.652863             1000   
1  16-02-2019   10:55:44           47.964508              100   
2  16-02-2019   12:10:24           51.697929              100   

   Current Range[A]  Sampling Pitch[uS]       Sensor  \
0                30                  10  Pyranometer   
1                10                  10  Pyranometer   
2                10                  10  Pyranometer   

   Caribration Constant (Er cal)[mV/kW/m^2]  PV Device Area[m^2]  \
0                                     52.49                    1   
1                                     52.49                    1   
2                                     52.49                    1   

   Current Temp. Coef.[A/degC]      ...        Voc (stc)[V]  Pm (stc)[W]  \
0                            0      ...          901.289707  5245.071996   
1                            0      ...           36.603990   389.943103   
2                            0      ...           36.628285   214.668875   

   Ipm (stc)[A]  Vpm (stc)[V]  Eta (stc)[%]   FF(stc)  Unnamed: 33  \
0      7.371913    711.493992     11.583047  0.692700          NaN   
1     10.372680     37.593285     17.987082  0.652611          NaN   
2      5.703425     37.638593     17.027440  0.647208          NaN   

   Voltage[V]         Module No.  Test-B result  
0  Current[A]  20190207125454041     defect_S_C  
1  Current[A]  20190216105544030       defect_C  
2  Current[A]  20190216121024130       defect_C  

[3 rows x 37 columns]

In [6]:
#drop some columns
dataWithClass.drop(["Date","Time","Voltage[V]","Module No."], axis=1, inplace=True)
print(dataWithClass.shape)
dataWithClass.head(2)

(196, 33)


Temperature[degC]  Voltge Range[V]  Current Range[A]  Sampling Pitch[uS]  \
0          47.652863             1000                30                  10   
1          47.964508              100                10                  10   

        Sensor  Caribration Constant (Er cal)[mV/kW/m^2]  PV Device Area[m^2]  \
0  Pyranometer                                     52.49                    1   
1  Pyranometer                                     52.49                    1   

   Current Temp. Coef.[A/degC]  Voltage Temp. Coef.[V/degC]  \
0                            0                            0   
1                            0                            0   

   Series Resistance(Rs)[ohm]      ...        T (stc)[degC]  Isc (stc)[A]  \
0                           0      ...                   25      8.194866   
1                           0      ...                   25      7.786469   

   Voc (stc)[V]  Pm (stc)[W]  Ipm (stc)[A]  Vpm (stc)[V]  Eta (stc)[%]  \
0    901.289707  5245.071996      7.371913    711.493992     11.583047   
1     36.603990   389.943103     10.372680     37.593285     17.987082   

    FF(stc)  Unnamed: 33  Test-B result  
0  0.692700          NaN     defect_S_C  
1  0.652611          NaN       defect_C  

[2 rows x 33 columns]

In [7]:
#drop na column
dataWithClass.dropna(inplace=True, axis=1)
print(dataWithClass.shape)
dataWithClass.head(2)

(196, 32)


Temperature[degC]  Voltge Range[V]  Current Range[A]  Sampling Pitch[uS]  \
0          47.652863             1000                30                  10   
1          47.964508              100                10                  10   

        Sensor  Caribration Constant (Er cal)[mV/kW/m^2]  PV Device Area[m^2]  \
0  Pyranometer                                     52.49                    1   
1  Pyranometer                                     52.49                    1   

   Current Temp. Coef.[A/degC]  Voltage Temp. Coef.[V/degC]  \
0                            0                            0   
1                            0                            0   

   Series Resistance(Rs)[ohm]      ...        Er (stc)[W/m^2]  T (stc)[degC]  \
0                           0      ...                   1000             25   
1                           0      ...                   1000             25   

   Isc (stc)[A]  Voc (stc)[V]  Pm (stc)[W]  Ipm (stc)[A]  Vpm (stc)[V]  \
0      8.194866    901.289707  5245.071996      7.371913    711.493992   
1      7.786469     36.603990   389.943103     10.372680     37.593285   

   Eta (stc)[%]   FF(stc)  Test-B result  
0     11.583047  0.692700     defect_S_C  
1     17.987082  0.652611       defect_C  

[2 rows x 32 columns]

In [8]:
#select object column
dataWithClass.select_dtypes(include = ['object']).head(3)

Sensor Test-B result
0  Pyranometer    defect_S_C
1  Pyranometer      defect_C
2  Pyranometer      defect_C

In [9]:
#Count number of data point in class/target
dataWithClass["Test-B result"].value_counts()

normal        97
defect_C      41
defect_P      39
defect_S_C    11
defect_S       5
defect_S_P     3
Name: Test-B result, dtype: int64

In [10]:
#create X(features) and y(class/target) data
X=dataWithClass.drop(['Sensor','Test-B result'],axis=1)
y = dataWithClass["Test-B result"]

In [11]:
#Shape of X and y
X.shape, y.shape

((196, 30), (196,))

In [12]:
# splinting data into X_train, X_test, y_train, y_test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.4,random_state=42, shuffle=False)

In [13]:
#number of data class labels
print("Train:", y_train.value_counts())

print("Test:", y_test.value_counts())

Train: normal        58
defect_P      25
defect_C      24
defect_S_C     6
defect_S_P     2
defect_S       2
Name: Test-B result, dtype: int64
Test: normal        39
defect_C      17
defect_P      14
defect_S_C     5
defect_S       3
defect_S_P     1
Name: Test-B result, dtype: int64


# ML Model

**ML algorithm selection:** 

1. A first approach to deal with imbalanced data is to balance it by discarding the majority class before applying an ML algorithm. The disadvantage of undersampling is that a model trained in this way will not perform well on real-world skewed test data since almost all the information was discarded. 

2. A second approach might be to oversample the minority class, say by the synthetic minority oversampling technique (SMOTE) contained in the 'imblearn' library.

3. I find, however, that the best result is obtained on the original dataset by using a ML algorithm based on ensembles of decision trees that intrinsically performs well on imbalanced data. Such algorithms not only allow for constructing a model that can cope with the missing values in our data, but they naturally allow for speedup via parallel-processing. Among these algorithms, XGBoost and random-forest algorithm used below slightly outperforms. Finally, XGBoost, RF, like several other ML algorithms, allows for weighting the positive class more compared to the negative class.

## With Original Data

### Random Forest

In [307]:
#Hyperparameter tuning
from sklearn.model_selection import RandomizedSearchCV, GridSearchCV
rfc=RandomForestClassifier(random_state=42)
param_grid = { 
    'n_estimators': [100, 200, 500],
    'max_features': ['auto', 'sqrt', 'log2'],
    'max_depth' : [4,5,6,7,8],
    'criterion' :['gini', 'entropy']
}
CV_rfc = GridSearchCV(estimator=rfc, param_grid=param_grid, cv= 2)
CV_rfc.fit(X_train, y_train)

GridSearchCV(cv=2, estimator=RandomForestClassifier(random_state=42),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [4, 5, 6, 7, 8],
                         'max_features': ['auto', 'sqrt', 'log2'],
                         'n_estimators': [100, 200, 500]})

In [308]:
#Best param
CV_rfc.best_params_

{'criterion': 'entropy',
 'max_depth': 5,
 'max_features': 'auto',
 'n_estimators': 500}

In [309]:
%%time
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report
r_clf=RandomForestClassifier(criterion='entropy',
 max_depth=5,
 max_features='auto',
 n_estimators=500)
r_clf.fit(X_train,y_train)
y_pred=r_clf.predict(X_test)
print("accuracy_score")
print(accuracy_score(y_test,y_pred))

print("classification_report")
print(classification_report(y_test,y_pred))

accuracy_score
0.8227848101265823
classification_report
              precision    recall  f1-score   support

    defect_C       0.94      0.94      0.94        17
    defect_P       0.73      0.79      0.76        14
    defect_S       0.00      0.00      0.00         3
  defect_S_C       0.00      0.00      0.00         5
  defect_S_P       0.00      0.00      0.00         1
      normal       0.81      0.97      0.88        39

    accuracy                           0.82        79
   macro avg       0.41      0.45      0.43        79
weighted avg       0.73      0.82      0.77        79

Wall time: 778 ms


C:\Users\Triloki-Gupta\Anaconda3\envs\envpytorch\lib\site-packages\sklearn\metrics\_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


### XGBoost

In [353]:
col = [str(i) for i in range(30)]
X=dataWithClass.drop(['Sensor','Test-B result'],axis=1)
y = dataWithClass["Test-B result"]
X.columns = col

In [355]:
# X.rename(columns=lambda x: x[0:5], inplace=True)
# splinting data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.4,random_state=42, shuffle=False)

In [357]:
from sklearn.model_selection import RandomizedSearchCV, GridSearchCV
xgb=XGBClassifier()
params = {
        'min_child_weight': [1, 5, 10],
        'gamma': [0.5, 1, 1.5, 2, 5],
        'subsample': [0.6, 0.8, 1.0],
        'colsample_bytree': [0.6, 0.8, 1.0],
        'max_depth': [3, 4, 5],
        'learning_rate':[0.01,0.1,1,10]
        }
CV_xgb = GridSearchCV(estimator=xgb, param_grid=params, cv= 3)
CV_xgb.fit(X, y)

GridSearchCV(cv=3,
             estimator=XGBClassifier(base_score=None, booster=None,
                                     colsample_bylevel=None,
                                     colsample_bynode=None,
                                     colsample_bytree=None, gamma=None,
                                     gpu_id=None, importance_type='gain',
                                     interaction_constraints=None,
                                     learning_rate=None, max_delta_step=None,
                                     max_depth=None, min_child_weight=None,
                                     missing=nan, monotone_constraints=None,
                                     n_estimators=100, n_jobs=None,
                                     num_parallel_tree=None, random_state=None,
                                     reg_alpha=None, reg_lambda=None,
                                     scale_pos_weight=None, subsample=None,
                                     tree_method=None,

In [358]:
CV_xgb.best_params_

{'colsample_bytree': 0.8,
 'gamma': 2,
 'learning_rate': 0.01,
 'max_depth': 3,
 'min_child_weight': 1,
 'subsample': 0.8}

In [367]:
%%time
from xgboost.sklearn import XGBClassifier
xgbc = XGBClassifier(colsample_bytree=0.8,gamma=2,learning_rate=0.01,max_depth=3,min_child_weight=1,subsample=0.8)
xgbc.fit(X_train, y_train)
y_pred=xgbc.predict(X_test)
print("accuracy_score")
print(accuracy_score(y_test,y_pred))

print("classification_report")
print(classification_report(y_test,y_pred))

accuracy_score
0.810126582278481
classification_report
              precision    recall  f1-score   support

    defect_C       0.94      0.94      0.94        17
    defect_P       0.71      0.71      0.71        14
    defect_S       0.00      0.00      0.00         3
  defect_S_C       0.00      0.00      0.00         5
  defect_S_P       0.00      0.00      0.00         1
      normal       0.81      0.97      0.88        39

    accuracy                           0.81        79
   macro avg       0.41      0.44      0.42        79
weighted avg       0.73      0.81      0.77        79

Wall time: 228 ms


## Up-sampling

### With all dataset

In [57]:
from imblearn.over_sampling import SMOTE
X=dataWithClass.drop(['Sensor','Test-B result'],axis=1)
y = dataWithClass["Test-B result"]

smote = SMOTE(k=2)
X_sm, y_sm = smote.fit_sample(X, y)
print(X_sm.shape, y_sm.shape)
X_train, X_test, y_train, y_test = train_test_split(X_sm, y_sm, test_size = 0.4,random_state=42)


(582, 30) (582,)


C:\Users\Triloki-Gupta\Anaconda3\lib\site-packages\imblearn\utils\deprecation.py:50: DeprecationWarning: 'k' is deprecated from 0.2 and will be removed in 0.4. Use 'k_neighbors' instead.
  category=DeprecationWarning)


In [58]:
%%time
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report
r_clf=RandomForestClassifier(criterion='entropy',
 max_depth=5,
 max_features='auto',
 n_estimators=500)
r_clf.fit(X_train,y_train)
y_pred=r_clf.predict(X_test)
print("accuracy_score")
print(accuracy_score(y_test,y_pred))

print("classification_report")
print(classification_report(y_test,y_pred))

accuracy_score
0.9570815450643777
classification_report
              precision    recall  f1-score   support

    defect_C       0.92      0.97      0.95        36
    defect_P       1.00      0.91      0.95        43
    defect_S       0.90      0.95      0.92        37
  defect_S_C       0.95      0.95      0.95        38
  defect_S_P       1.00      1.00      1.00        39
      normal       0.97      0.97      0.97        40

   micro avg       0.96      0.96      0.96       233
   macro avg       0.96      0.96      0.96       233
weighted avg       0.96      0.96      0.96       233

Wall time: 1.51 s


In [59]:
%%time
from xgboost.sklearn import XGBClassifier
xgbc = XGBClassifier(colsample_bytree=0.8,gamma=2,learning_rate=0.01,max_depth=3,min_child_weight=1,subsample=0.8)
xgbc.fit(X_train, y_train)
y_pred=xgbc.predict(X_test)
print("accuracy_score")
print(accuracy_score(y_test,y_pred))

print("classification_report")
print(classification_report(y_test,y_pred))

accuracy_score
0.927038626609442
classification_report
              precision    recall  f1-score   support

    defect_C       0.92      0.97      0.95        36
    defect_P       1.00      0.77      0.87        43
    defect_S       0.88      0.95      0.91        37
  defect_S_C       0.88      0.95      0.91        38
  defect_S_P       0.95      1.00      0.97        39
      normal       0.95      0.95      0.95        40

   micro avg       0.93      0.93      0.93       233
   macro avg       0.93      0.93      0.93       233
weighted avg       0.93      0.93      0.93       233

Wall time: 825 ms


### Within train data only

In [96]:
# col = [str(i) for i in range(30)]
X=dataWithClass.drop(['Sensor','Test-B result'],axis=1)
y = dataWithClass["Test-B result"]
# X.columns = col

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.4,random_state=42,shuffle= False)

smote = SMOTE(k=1)
X_sm, y_sm = smote.fit_sample(X_train, y_train)
print(X_sm.shape, y_sm.shape)
# X_test.columns = X_sm.columns()

(348, 30) (348,)


C:\Users\Triloki-Gupta\Anaconda3\lib\site-packages\imblearn\utils\deprecation.py:50: DeprecationWarning: 'k' is deprecated from 0.2 and will be removed in 0.4. Use 'k_neighbors' instead.
  category=DeprecationWarning)


In [97]:
from sklearn.model_selection import RandomizedSearchCV, GridSearchCV
rfc=RandomForestClassifier(random_state=42)
param_grid = { 
    'n_estimators': [100, 200, 500],
    'max_features': ['auto', 'sqrt', 'log2'],
    'max_depth' : [4,5,6,7,8],
    'criterion' :['gini', 'entropy']
}
CV_rfc = GridSearchCV(estimator=rfc, param_grid=param_grid, cv= 2)
CV_rfc.fit(X_sm, y_sm)

GridSearchCV(cv=2, error_score='raise-deprecating',
       estimator=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators='warn', n_jobs=None,
            oob_score=False, random_state=42, verbose=0, warm_start=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'n_estimators': [100, 200, 500], 'max_features': ['auto', 'sqrt', 'log2'], 'max_depth': [4, 5, 6, 7, 8], 'criterion': ['gini', 'entropy']},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [98]:
CV_rfc.best_params_

{'criterion': 'gini',
 'max_depth': 6,
 'max_features': 'log2',
 'n_estimators': 100}

In [114]:
%%time
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report
r_clf=RandomForestClassifier(criterion='gini',
 max_depth=6,
 max_features='log2',
 n_estimators=100,random_state=42)
r_clf.fit(X_sm,y_sm)
y_pred=r_clf.predict(X_test)
print("accuracy_score")
print(accuracy_score(y_test,y_pred))

print("classification_report")
print(classification_report(y_test,y_pred))

accuracy_score
0.810126582278481
classification_report
              precision    recall  f1-score   support

    defect_C       0.94      0.94      0.94        17
    defect_P       1.00      0.36      0.53        14
    defect_S       0.00      0.00      0.00         3
  defect_S_C       0.31      0.80      0.44         5
  defect_S_P       0.33      1.00      0.50         1
      normal       0.93      0.97      0.95        39

   micro avg       0.81      0.81      0.81        79
   macro avg       0.58      0.68      0.56        79
weighted avg       0.86      0.81      0.80        79

Wall time: 175 ms


C:\Users\Triloki-Gupta\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


In [93]:
from sklearn.model_selection import RandomizedSearchCV, GridSearchCV
xgb=XGBClassifier()
params = {
        'min_child_weight': [1, 5, 10],
        'gamma': [0.5, 1, 1.5, 2, 5],
        'subsample': [0.6, 0.8, 1.0],
        'colsample_bytree': [0.6, 0.8, 1.0],
        'max_depth': [3, 4, 5],
        'learning_rate':[0.01,0.1,1,10]
        }
CV_xgb = GridSearchCV(estimator=xgb, param_grid=params, cv= 3)
CV_xgb.fit(X_sm, y_sm)
CV_xgb.best_params_

C:\Users\Triloki-Gupta\Anaconda3\lib\site-packages\sklearn\model_selection\_search.py:841: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


{'colsample_bytree': 0.6,
 'gamma': 0.5,
 'learning_rate': 0.01,
 'max_depth': 4,
 'min_child_weight': 1,
 'subsample': 1.0}

In [95]:
%%time
pd.DataFrame(X_test).columns = ['f0', 'f1', 'f2', 'f3', 'f4', 'f5', 'f6', 'f7', 'f8', 'f9', 'f10', 'f11', 'f12', 'f13', 'f14', 'f15', 'f16', 'f17', 'f18', 'f19', 'f20', 'f21', 'f22', 'f23', 'f24', 'f25', 'f26', 'f27', 'f28', 'f29']
from xgboost.sklearn import XGBClassifier
xgbc = XGBClassifier(n_estimators=200,colsample_bytree=0.6,gamma=0.5,learning_rate=0.01,max_depth=4,min_child_weight=1,subsample=1.0)
xgbc.fit(X_sm, y_sm)
y_pred=xgbc.predict(X_test)
print("accuracy_score")
print(accuracy_score(y_test,y_pred))

print("classification_report")
print(classification_report(y_test,y_pred))

accuracy_score
0.8354430379746836
classification_report
              precision    recall  f1-score   support

    defect_C       0.94      0.94      0.94        17
    defect_P       0.90      0.64      0.75        14
    defect_S       0.00      0.00      0.00         3
  defect_S_C       0.67      0.40      0.50         5
  defect_S_P       0.50      1.00      0.67         1
      normal       0.81      0.97      0.88        39

   micro avg       0.84      0.84      0.84        79
   macro avg       0.64      0.66      0.62        79
weighted avg       0.81      0.84      0.81        79

Wall time: 1.12 s


C:\Users\Triloki-Gupta\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


## Conclusion

Thoroughly interrogated the data with the help of EDA.

To deal with the large skew in the data, we chose an appropriate metric and used an ML algorithm based on an ensemble of decision trees which works best with strongly imbalanced classes.

**1. With original data**

- Random Forest: 82.27%
- XGBoost: 81.01%

**2. Upsampling(SMOTE)**

**- Within all data**

- Random Forest: 95.70%
- XGBoost: 92.7%

**- Within train data only**

- Random Forest: 81.01%
- XGBoost: 83.54%

Less number of data point for defect defect_S_C, defect_S & defect_S_P which is impacting accuracy.

- defect_S_C:    11
- defect_S:       5
- defect_S_P:     3